gh-68114: Fix handling for removed PyArg_ParseTuple formatters

llllllllll opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->
re-opening this as a github pull request @serhiy-storchaka 

The only difference from the old patch is that I used `blurb` to write the news entry.

<!-- issue-number: bpo-23926 -->
https://bugs.python.org/issue23926
<!-- /issue-number -->


<!-- gh-issue-number: gh-68114 -->
* Issue: gh-68114
<!-- /gh-issue-number -->

Commit:
Fix handling for removed PyArg_ParseTuple formatters

vstinner commented on Python/getargs.c:
```diff
@@ -2303,6 +2303,14 @@ skipitem(const char **p_format, va_list *p_va, int flags)
 
     /* string codes */
 
+    case 'w': /* buffer, read-write */
```
Would you mind to move back this block to line 2323 to make the change simpler to follow/review? Here it seems like you are adding *new* code, rather than modifying existing code.

llllllllll replied:
This actually is new code, not just a change. Moving it would alter the functionality due to the way the fall through works.

encukou replied:
Unfortunately it's missing the `va_arg` call from the code it's replacing, so it fails to skip `w*` properly.

ZackerySpytz commented on Modules/_testcapimodule.c:
```diff
@@ -1437,6 +1437,77 @@ getargs_Z(PyObject *self, PyObject *args)
+        if (PyArg_ParseTupleAndKeywords(args, kwargs, formats_3[n],
+                                        (char**) keywords,
+                                        &tmp, &tmp, &tmp)) {
+            Py_DECREF(args);
```
I believe `kwargs` should be decrefed here as well.

ZackerySpytz commented on Modules/_testcapimodule.c:
```diff
@@ -1437,6 +1437,77 @@ getargs_Z(PyObject *self, PyObject *args)
+        if (PyArg_ParseTupleAndKeywords(args, kwargs, formats_4[n],
+                                        (char**) keywords,
+                                        &tmp, &tmp, &tmp, &tmp)) {
+            Py_DECREF(args);
```
I believe `kwargs` should be decrefed here as well.

ZackerySpytz commented on Modules/_testcapimodule.c:
```diff
@@ -1437,6 +1437,77 @@ getargs_Z(PyObject *self, PyObject *args)
+        return NULL;
+    }
+
+    for (n = 0;formats_3[n];++n) {
```
```suggestion
    for (n = 0; formats_3[n]; ++n) {
```

ZackerySpytz commented on Modules/_testcapimodule.c:
```diff
@@ -1437,6 +1437,77 @@ getargs_Z(PyObject *self, PyObject *args)
+        return NULL;
+    }
+
+    for (n = 0;formats_4[n];++n) {
```
```suggestion
    for (n = 0; formats_4[n]; ++n) {
```

ZackerySpytz commented on Modules/_testcapimodule.c:
```diff
@@ -1437,6 +1437,77 @@ getargs_Z(PyObject *self, PyObject *args)
+        return NULL;
+    }
+
+    if (!(kwargs = PyDict_New()) || PyDict_SetItemString(kwargs, "c", Py_None)) {
```
Please limit lines to 79 chars.

MaxwellDupre reviewed (approved):
Test_capi produces:
Ran 40 tests in 7.553s
OK
Looks ok.

iritkatriel reviewed (dismissed):
This has merge conflicts now.

encukou committed:
Merge in the main branch

encukou committed:
Merge comments

encukou committed:
Recreate the blurb

encukou committed:
Fix the test

- Don't use removed trivial helper raiseTestError
- Decref kwargs if it's created
- Export the test so that it is run

encukou committed:
Fix style

encukou committed:
Ensure the invalid args result in SystemError

encukou committed:
Fix functionality, and add additional test

The original patch didn't call va_arg, resulting in corruption
when 'w*' was used.

encukou committed:
Unmark 'w#' as supported in test_skipitem_with_suffix

encukou committed:
Don't omit parameter names

encukou committed:
Merge branch 'main' into remove-old-formatters
